### PR TITLE
RavenDB-9373 & RavenDB-9475

### PR DESCRIPTION
**Pull request by karmeli87**
RavenDB-9475 Add the option to remove cmpxchng values
RavenDB-9373 Add endpoint for cmpxchng values

### CODE REVIEW

File: src/Raven.Client/Documents/Operations/CompareExchange.cs [hunk 1]
```diff
@@ -56,6 +56,143 @@ public static CmpXchgResult<T> ParseFromBlittable(BlittableJsonReaderObject resp
...
+            }
+
+            public override bool IsReadRequest => true;
+            public override HttpRequestMessage CreateRequest(JsonOperationContext ctx, ServerNode node, out string url)
+            {
+                url = $"{node.Url}/databases/{node.Database}/cmpxchg/list?key={_keyPrefix}";
+                
+                if (_page != null)
+                {
+                    url += $"&page={_page}";
```
Comment (ayende) [inline src/Raven.Client/Documents/Operations/CompareExchange.cs:100]: use `start` for consistency

File: src/Raven.Client/Documents/Operations/CompareExchange.cs [hunk 2]
```diff
@@ -56,6 +56,143 @@ public static CmpXchgResult<T> ParseFromBlittable(BlittableJsonReaderObject resp
...
+                }
+                
+                if (_size != null)
+                {
+                    url += "&size={_size}";
```
Comment (ayende) [inline src/Raven.Client/Documents/Operations/CompareExchange.cs:105]: use `pageSize` for consistency

File: src/Raven.Client/Documents/Operations/CompareExchange.cs [hunk 3]
```diff
@@ -56,6 +56,143 @@ public static CmpXchgResult<T> ParseFromBlittable(BlittableJsonReaderObject resp
...
+                _keyPrefix = keyPrefix;
+                _page = page;
+                _size = size;
+                _conventions = conventions;
+            }
+
+            public override bool IsReadRequest => true;
+            public override HttpRequestMessage CreateRequest(JsonOperationContext ctx, ServerNode node, out string url)
+            {
+                url = $"{node.Url}/databases/{node.Database}/cmpxchg/list?key={_keyPrefix}";
```
Comment (ayende) [inline src/Raven.Client/Documents/Operations/CompareExchange.cs:96]: use `startsWith` for consistency

File: src/Raven.Server/ServerWide/ClusterStateMachine.cs [hunk 4]
```diff
@@ -972,22 +956,58 @@ public unsafe void CompareExchange(TransactionOperationContext context, string t
...
+                    
+                    if(pageSize == 0)
+                        yield break;
+                }
+            }
+        }
+
+        private static unsafe string ReadCmpXchgKey(TransactionOperationContext context, TableValueReader reader)
+        {
+            return new LazyStringValue(null, reader.Read((int)UniqueItems.Key, out var size), size, context);
```
Comment (ayende) [inline src/Raven.Server/ServerWide/ClusterStateMachine.cs:999]: Do you really have to allocate a string here?

File: src/Raven.Server/ServerWide/Commands/CompareExchangeCommands.cs [hunk 5]
```diff
@@ -0,0 +1,112 @@
...
+        {
+            var dbKey = Key.ToLowerInvariant();
+            BlittableJsonReaderObject result = null;
+            using (Slice.From(context.Allocator, dbKey, out Slice keySlice))
+            {
+                if (items.ReadByKey(keySlice, out var reader))
+                {
+                    var itemIndex = *(long*)reader.Read((int)ClusterStateMachine.UniqueItems.Index, out var _);
+                    var storeValue = reader.Read((int)ClusterStateMachine.UniqueItems.Value, out var size);
+                    result = new BlittableJsonReaderObject(storeValue, size, context);
```
Comment (ayende) [inline src/Raven.Server/ServerWide/Commands/CompareExchangeCommands.cs:55]: This is not a good idea. You just deleted this object, and now you are returning an pointer to it?
This is now pointing to undefined memory. You need to clone it.

File: src/Raven.Server/Web/System/CompareExchangeHandler.cs [hunk 6]
```diff
@@ -70,5 +71,69 @@ public Task GetCmpXchgValue()
...
+            using (ServerStore.ContextPool.AllocateOperationContext(out TransactionOperationContext context))
+            {
+                var command = new RemoveCompareExchangeCommand(key, index);
+                using (var writer = new BlittableJsonTextWriter(context, ResponseBodyStream()))
+                {
+                    (var raftIndex, var result) = await ServerStore.SendToLeaderAsync(command);
+                    await ServerStore.Cluster.WaitForIndexNotification(raftIndex);
+                    using (context.OpenReadTransaction())
+                    {
+                        var res = ServerStore.Cluster.GetCmpXchg(context, key);
```
Comment (ayende) [inline src/Raven.Server/Web/System/CompareExchangeHandler.cs:95]: Why are you doing this? This is invalid.
You are getting this from a totally different value.
You need to use `raftIndex`

File: src/Raven.Server/Web/System/CompareExchangeHandler.cs [hunk 7]
```diff
@@ -70,5 +71,69 @@ public Task GetCmpXchgValue()
...
+            }
+        }
+
+        [RavenAction("/databases/*/cmpxchg/list", "GET", AuthorizationStatus.ValidUser)]
+        public Task ListCmpXchgValues()
+        {
+            var prefix = Database.Name + "/";
+            var key = prefix + GetStringQueryString("key", false);
+            var page = GetIntValueQueryString("page", false) ?? 0;
+            var size = GetIntValueQueryString("size", false) ?? 1024;
```
Comment (ayende) [inline src/Raven.Server/Web/System/CompareExchangeHandler.cs:114]: These parameters don't follow our standard conventions. Use `GetStart`, `GetPageSize`, etc.

File: src/Raven.Server/Web/System/CompareExchangeHandler.cs (hunk 6 above)
Comment (karmeli87) [reply src/Raven.Server/Web/System/CompareExchangeHandler.cs:95]: We try to delete the compare exchange value here, so if we successful (or the value hasn't existed at the first place) then res will contain (-1, null).

**Pull request merged**